crates/matrix-sdk-ui/src/unable_to_decrypt_hook.rs

```diff
@@ -72,35 +76,32 @@ pub struct UtdHookManager {
     /// The parent hook we'll call, when we have found a unique UTD.
     parent: Arc<dyn UnableToDecryptHook>,
 
-    /// A mapping of events we've marked as UTDs, and the time at which we
-    /// observed those UTDs.
-    ///
-    /// Note: this is unbounded, because we have absolutely no idea how long it
-    /// will take for a UTD to resolve, or if it will even resolve at any
-    /// point.
-    known_utds: Arc<Mutex<HashMap<OwnedEventId, Instant>>>,
-
     /// An optional delay before marking the event as UTD ("grace period").
     max_delay: Option<Duration>,
 
-    /// The set of outstanding tasks to report deferred UTDs, including the
-    /// event relating to the task.
+    /// A mapping of events we're going to report as UTDs, to the tasks to do
+    /// so.
     ///
     /// Note: this is empty if no [`Self::max_delay`] is set.
     ///
     /// Note: this is theoretically unbounded in size, although this set of
     /// tasks will degrow over time, as tasks expire after the max delay.
-    pending_delayed: Arc<Mutex<PendingUtdReports>>,
+    pending_delayed: Arc<Mutex<HashMap<OwnedEventId, PendingUtdReport>>>,
+
+    /// The list of events we've reported as UTDs.
+    ///
+    /// Note: this is unbounded, because we could receive infinitely many UTDs.
+    reported_utds: Arc<Mutex<HashSet<OwnedEventId>>>,
```

Comment:
I think I slightly disagree with this change, because:

1. it serves the same interest as `known_utds`, aka avoiding double-reporting.
2. in contrary to `known_utds`, it only grows over time, while `known_utds` would shrink, keeping it under a reasonable size limit.

So, instead of the change in the last commit, can we make `known_utds` a set instead of a map, and keep the same inverted semantics that was there at the beginning?

Reply:
Well, I *can*, but on the other hand, I need it to be `reported_utds` rather than `known_utds` for my next change, which will be a fix for https://github.com/matrix-org/matrix-rust-sdk/issues/3374, and in which we will persist this data to disk. [The intention is to use a Bloom filter to reduce the storage requirement of this.]  When reloading the data into future instances of `UtdHookManager`, the question we need to know is not "had we seen this event as a UTD", but "had we seen this event as a UTD *and then reported it*".

So, given that, can we leave it as is?

Reply:
Are you saying that the database will serve as the unique source of truth, and there won't be the in-memory storing of already-reported events (aka you're getting rid of `reported_utds` later? If so, yes that's fine to keep as is. Otherwise, that means that this in-memory will grow, independently of what we have in storage, so I'd like to try our best to keep an upper bound on it, and invert the semantics as I've suggested before.

Reply:
> Are you saying that the database will serve as the unique source of truth, and there won't be the in-memory storing of already-reported events (aka you're getting rid of reported_utds later?)

Not really, no. The intention is to keep this data (though in a much more compact form) in memory. It will be initialised from the database, and changes will be persisted promptly, though not immediately. Loading the data from the store isn't free, so we want to avoid doing it on every UTD, so we still need the in-memory copy.

To avoid back-and-forth here, I'll make the change you've asked for, and we can revisit further as part of the PR that fixes #3374.